src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -727,11 +744,159 @@ BoundExpression convertArrayElement(BoundExpression element, TypeSymbol elementT
                 collectionTypeKind: CollectionLiteralTypeKind.None,
                 placeholder: null,
                 collectionCreation: null,
+                collectionBuilderMethod: null,
                 elements: builder.ToImmutableAndFree(),
                 targetType,
                 hasErrors: true);
         }
 
+        private BoundCollectionLiteralExpression BindCollectionBuilderCollectionLiteral(
+            BoundUnconvertedCollectionLiteralExpression node,
+            NamedTypeSymbol targetType,
+            bool wasCompilerGenerated,
+            BindingDiagnosticBag diagnostics)
+        {
+            var syntax = (ExpressionSyntax)node.Syntax;
+
+            bool hasAttribute = targetType.HasCollectionBuilderAttribute(out TypeSymbol? builderType, out string? methodName);
+            Debug.Assert(hasAttribute);
+
+            TryGetCollectionIterationType(syntax, targetType.OriginalDefinition, out TypeWithAnnotations elementTypeOriginalDefinition);
+            if (!elementTypeOriginalDefinition.HasType)
```

Comment:
Yes, I believe so, but it seemed cleaner (and safer) to ignore the result of `TryGetCollectionIterationType` and simply check `HasType` unconditionally rather than asserting `HasType` when `TryGetCollectionIterationType` returns `true`.